usage/camp/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/creation/CampResolver.java

```diff
@@ -0,0 +1,144 @@
+        // should have a "services: [ ... ]" block (and which may subsequently be unwrapped)
+        BrooklynClassLoadingContext loader = CatalogUtils.newClassLoadingContext(mgmt, item, loaderO);
+
+        Set<String> encounteredTypes;
+        // symbolicName could be null if coming from the catalog parser where it tries to load before knowing the id
+        if (item.getSymbolicName() != null) {
+            encounteredTypes = ImmutableSet.<String>builder()
+                .addAll(parentEncounteredTypes)
+                .add(item.getSymbolicName())
+                .build();
+        } else {
+            encounteredTypes = parentEncounteredTypes;
+        }
+
+        AbstractBrooklynObjectSpec<?, ?> spec;
+        String planYaml = RegisteredTypes.getImplementationDataStringForSpec(item);
+        MutableSet<Object> supers = MutableSet.copyOf(item.getSuperTypes());
+        supers.addIfNotNull(expectedType);
+        if (RegisteredTypes.isSubTypeOf(supers, Policy.class)) {
+            spec = CampInternalUtils.createPolicySpec(planYaml, loader, encounteredTypes);
+        } else if (RegisteredTypes.isSubTypeOf(supers, Location.class)) {
+            spec = CampInternalUtils.createLocationSpec(planYaml, loader, encounteredTypes);
+        } else if (RegisteredTypes.isSubTypeOf(supers, Application.class)) {
+            spec = createEntitySpecFromServicesBlock(planYaml, loader, encounteredTypes, true);
+        } else if (RegisteredTypes.isSubTypeOf(supers, Entity.class)) {
+            spec = createEntitySpecFromServicesBlock(planYaml, loader, encounteredTypes, false);
+        } else {
+            // try any of them???
+            
+            throw new IllegalStateException("Cannot detect spec type from "+item.getSuperTypes()+" for "+item+"\n"+planYaml);
+        }
+
+        ((AbstractBrooklynObjectSpec<?, ?>)spec).catalogItemId(item.getId());
+
+        if (Strings.isBlank( ((AbstractBrooklynObjectSpec<?, ?>)spec).getDisplayName() ))
+            ((AbstractBrooklynObjectSpec<?, ?>)spec).displayName(item.getDisplayName());
+
+        return spec;
+    }
+ 
+    private static EntitySpec<?> createEntitySpecFromServicesBlock(String plan, BrooklynClassLoadingContext loader, Set<String> encounteredTypes, boolean isApplication) {
+        CampPlatform camp = CampInternalUtils.getCampPlatform(loader.getManagementContext());
+
+        AssemblyTemplate at = CampInternalUtils.registerDeploymentPlan(plan, loader, camp);
+        AssemblyTemplateInstantiator instantiator = CampInternalUtils.getInstantiator(at);
+        if (instantiator instanceof AssemblyTemplateSpecInstantiator) {
+            EntitySpec<? extends Application> appSpec = ((AssemblyTemplateSpecInstantiator)instantiator).createApplicationSpec(at, camp, loader, encounteredTypes);
+
+            if (!isApplication && EntityManagementUtils.canPromoteChildrenInWrappedApplication(appSpec) && appSpec.getChildren().size()==1) {
```

Comment:
`appSpec.getChildren().size()==1` already included in the `canPromoteChildrenInWrappedApplication` check.
Also (minor) can move this block in a wrapper function instead of using a flag argument.